sentinel-core/src/main/java/com/alibaba/csp/sentinel/spi/SpiLoader.java

```diff
@@ -0,0 +1,525 @@
+                    @Override
+                    public int compare(Class<? extends S> o1, Class<? extends S> o2) {
+                        Spi spi1 = o1.getAnnotation(Spi.class);
+                        int order1 = spi1 == null ? 0 : spi1.order();
+
+                        Spi spi2 = o2.getAnnotation(Spi.class);
+                        int order2 = spi2 == null ? 0 : spi2.order();
+
+                        return Integer.compare(order1, order2);
+                    }
+                });
+            } catch (IOException e) {
+                fail("error reading SPI configuration file", e);
+            } finally {
+                closeResources(in, br);
+            }
+        }
+    }
+
+    @Override
+    public String toString() {
+        return "com.alibaba.csp.sentinel.spi.SpiLoader[" + service.getName() + "]";
+    }
+
+    /**
+     * Create Provider instance list
+     *
+     * @param clazzList class types of Providers
+     * @return Provider instance list
+     */
+    private List<S> createInstanceList(List<Class<? extends S>> clazzList) {
+        if (clazzList == null || clazzList.size() == 0) {
+            return Collections.emptyList();
+        }
+
+        List<S> instances = new ArrayList<>(clazzList.size());
+        for (Class<? extends S> clazz : clazzList) {
+            S instance = createInstance(clazz);
+            instances.add(instance);
+        }
+        return instances;
+    }
+
+    /**
+     * Create Provider instance
+     *
+     * @param clazz class type of Provider
+     * @return Provider class
+     */
+    private S createInstance(Class<? extends S> clazz) {
+        Spi spi = clazz.getAnnotation(Spi.class);
+        boolean singleton = true;
+        if (spi != null) {
+            singleton = spi.isSingleton();
+        }
+        return createInstance(clazz, singleton);
+    }
+
+    /**
+     * Create Provider instance
+     *
+     * @param clazz     class type of Provider
+     * @param singleton if instance is singleton or prototype
+     * @return Provider instance
+     */
+    private S createInstance(Class<? extends S> clazz, boolean singleton) {
+        S instance = null;
+        try {
+            if (singleton) {
+                instance = singletonMap.get(clazz.getName());
+                if (instance == null) {
+                    synchronized (this) {
+                        instance = service.cast(clazz.newInstance());
+                        singletonMap.put(clazz.getName(), instance);
+                    }
+                }
+            } else {
+                instance = service.cast(clazz.newInstance());
+            }
+        } catch (Throwable e) {
+            fail(clazz.getName() + " could not be instantiated");
+        }
+        return instance;
+    }
+
+    /**
+     * Close all resources
+     *
+     * @param closeables {@link Closeable} resources
+     */
+    private void closeResources(Closeable... closeables) {
```

Comment:
here i find usage only close one stream one time, so this may change for close one.
if u want to close more, i think `for` loop should catch each fail and finish all close task, and then return fail message.
want your opinion, plz.

Reply:
Now it has been changed, record the first exception during close resources, to make it can finish all close task, and then throw exception with fail message. Thanks for your careful review.